### PR TITLE
GDPR anonymization is much faster

### PR DESCRIPTION
**Pull request by matllubos**

### CODE REVIEW

File: gdpr/anonymizers/model_anonymizers.py [hunk 1]
```diff
@@ -342,29 +342,33 @@ def update_obj(self, obj: Model, legal_reason: Optional[LegalReason] = None,
...
-            versions_update_dict = [
-                (
-                    version,
-                    {
-                        name: self.get_value_from_version(self[name], obj, version, name, anonymization=anonymization)
-                        for name in raw_local_fields
-                        if name in get_reversion_local_field_dict(version)
-                    }
+
+        if raw_local_fields:
```
Comment (matllubos) [inline gdpr/anonymizers/model_anonymizers.py:346]: if there is nothing to anonymize no action is necessary.. For example all reversion data was get from DB but for nothing

File: gdpr/management/commands/deactivate_expired_reasons.py [hunk 2]
```diff
@@ -2,10 +2,16 @@
...
 from gdpr.models import LegalReason
 
+from tqdm import tqdm
+
+from django.db import transaction
+
 
 class Command(BaseCommand):
 
+    @transaction.atomic
```
Comment (matllubos) [inline gdpr/management/commands/deactivate_expired_reasons.py:12]: command should run in transaction

File: gdpr/models.py [hunk 3]
```diff
@@ -116,7 +116,7 @@ def filter_expired_retaining_data_in_last_days(self, days=None):
             'expires_at__lt': timezone.now()
         }
 
-        return self.filter(purpose_slug__in=purpose_slugs_retaining_data, **filter_keys)
+        return self.filter(is_active=True, purpose_slug__in=purpose_slugs_retaining_data, **filter_keys)
```
Comment (matllubos) [inline gdpr/models.py:119]: I think that only not active LegalReasons should be selected for anonymization

File: gdpr/purposes/default.py [hunk 4]
```diff
@@ -86,6 +86,10 @@ def anonymize_obj(self, obj: Model, legal_reason: Optional["LegalReason"] = None
                                set([i.purpose_slug for i in other_legal_reasons])]:
             parsed_fields -= allowed_fields
 
+        if len(parsed_fields) == 0:
+            # If there are no fields to anonymize do nothing.
+            return
```
Comment (matllubos) [inline gdpr/purposes/default.py:91]: the same situation as above: "no fields, no anonymization"

File: gdpr/utils.py [hunk 5]
```diff
@@ -56,7 +56,10 @@ def get_reversion_versions(obj: Any) -> QuerySet:
     if hasattr(Version.objects, 'get_for_object'):
         return Version.objects.get_for_object(obj).order_by('id')
```
Comment (matllubos) [inline gdpr/utils.py:57]: for new django revision it will be slow. There is no db index :(

File: gdpr/utils.py [hunk 6]
```diff
@@ -56,7 +56,10 @@ def get_reversion_versions(obj: Any) -> QuerySet:
...
     content_type = ContentType.objects.get_for_model(obj.__class__)
-    return Version.objects.filter(content_type=content_type, object_id=obj.pk).order_by('id')
+    if isinstance(obj.pk, int):
+        return Version.objects.filter(content_type=content_type, object_id_int=obj.pk).order_by('id')
+    else:
+        return Version.objects.filter(content_type=content_type, object_id=obj.pk).order_by('id')
```
Comment (matllubos) [inline gdpr/utils.py:62]: this is little bit tricky. Integer primary keys are stored to the object_id_int field which has db index

**Review by Formulka — APPROVED** [review]

File: gdpr/encryption.py [hunk 7]
```diff
@@ -168,7 +168,7 @@ def translate_email_address(key: str, email: str, encrypt: bool = True, restrict
 
     """
     local, domain_tld = email.split("@")
-    domain, tld = domain_tld.split(".")
+    domain, tld = domain_tld.rsplit(".", 1)
```
Comment (matllubos) [inline gdpr/encryption.py:171]: one more bug. Domain can contains more spots..

File: gdpr/utils.py (hunk 6 above)
Comment (fdobrovolny) [reply gdpr/utils.py:62]: I'm just thinking if there shouldn't be some check that the `object_id_int` exists. But the tests are passing so I'm guessing there is no need for that.

**Review by fdobrovolny — APPROVED** [review]
I had just one concern which is mentioned in the comments other than that LGTM.

File: gdpr/utils.py (hunk 6 above)
Comment (matllubos) [reply gdpr/utils.py:62]: it is not necessary, because the old version have it and the new version of reversion have method get_for_object

**Pull request merged**